docs/developer/decision-records/2025-07-25-multiple-participant-identifiers/README.md

```diff
@@ -0,0 +1,112 @@
+for a given protocol:
+
+```java
+@NotNull
+String getParticipantId(String protocol);
+```
+
+The previously used participant ID will remain the default and will be set in the `DataspaceProfileContextRegistry` as
+such. Thus, if no dedicated participant ID is defined for the given protocol, the method will return the default ID:
+
+```java
+public String getParticipantId(String protocol) {
+    return profiles().stream()
+            .filter(it -> it.name().equals(protocol))
+            .map(DataspaceProfileContext::participantId)
+            .findAny()
+            .orElse(defaultParticipantId);
+}
+```
+
+In all places where the participant ID is used (namely during creation of a catalog or a contract agreement), an
+instance of `DataspaceProfileContextRegistry` will be added, and the registry's `getParticipantId` method will be
+used where currently the participant ID is inserted.
+
+### Extraction of counter-party ID
+
+Right now, extraction of the counter-party ID from a `ClaimToken` is done in the `ParticipantAgentServiceImpl`, either
+through a `ParticipantAgentServiceExtension` or, as a fallback solution, by looking for a specific claim that can be
+defined through configuration.
+
+The extraction will be pulled up into the `ProtocolTokenValidatorImpl` to separate extraction of the ID from the
+extraction of any other participant attributes, which will still be handled by the `ParticipantAgentServiceExtensions`.
+The extracted ID will then be passed to the `ParticipantAgentService` as a parameter.
+
+As ID extraction should depend on the `DataspaceProfileContext`, an additional method will be added to the
+`DataspaceProfileContextRegistry` for registering a new profile. In addition to the profile itself, this method will
+also accept a function for extracting the participant ID from a `ClaimToken`, an instance of
+`Function<ClaimToken, String>`:
+
+```java
+void register(DataspaceProfileContext context, Function<ClaimToken, String> idExtractionFunction);
+```
+
+Accordingly, another method will be added to obtain the ID extraction function for a given profile context:
+
+```java
+@NotNull
+Function<ClaimToken, String> getIdExtractionFunction(String protocol);
+```
+
+All registered profiles that were created without passing a function for ID extraction will use the default function.
+This will be supplied during creation of the `DataspaceProfileContextRegistryImpl` and contain the current default
+behaviour of looking for a configurable claim in the `ClaimToken`. As the default logic may be different depending on
+the `iam` extension used (e.g. for DCP the credential subject ID from received VCs is used), the default function must
+be overridable. Therefore, a third method will be added to the `DataspaceProfileContextRegistry`: 
+
+```java
+void setDefaultIdExtractionFunction(Function<ClaimToken, String> extractionFunction);
```

Comment:
I think the default function should only be functional for the "out of the box" profile, useful for testing and samples, but for a production dataspace I foresee a custom profile definition in any case so the adopters will be in control of the profile and they wouldn't be surprised when the default EDC implementation change for whatever reason.

TL;DR: I don't think we need this set method

Reply:
you're right, different identity services have different ways to extract the identifier, and it's correct that there's at least one default way to do so... should this be described by a `DefaultIdExtractionFunction` that will be registered by either iam-mock and dcp implementations and it will be put on the default dataspace profile.